Avoid ES aggregation in a util that hits AppES

sravfeyn opened the pull request:
Fixes https://sentry.io/organizations/dimagi/issues/1529243347/?project=1545828&query=is%3Aunresolved

@orangejenny 
@snopoke

sravfeyn committed:
Avoid ES aggregation

snopoke reviewed (commented):
Do you understand how this 'field_data' can be 8.7gb? I think a better solution here would be to change the app index to make the 'lang' field `not_analyzed`. Looking at the mapping for the index I think most of the fields can be marked as `not_analyzed` except for a few like: `build_comment` `comment_from`, `description`.

That would require a reindex but would be a better long term solution.

sravfeyn commented:
This fielddata limit is not directly caused by the app index, as it is very small (the whole index size itself is not more than 1.5Gb). This limit is hit on the node which this index shares other indices with. 

Agreed about the comments on `not_analyzed` on app-mapping, though it wouldn't fix this particular issue since it is tied to other index usage of fielddata as well.

snopoke commented:
@sravfeyn I think it would fix this issue since it would not require field_data at all in order to perform the query right? According to the docs, field data is only needed for analyzed text fields.

orangejenny commented:
Tangential thought: this code is only used to populate the language dropdown on edit mobile workers. Would it make sense to also make that dropdown use ajax instead of pulling languages every time the page is loaded?

sravfeyn committed:
Revert "Avoid ES aggregation"

This reverts commit 294001bf8b2896c469646c8846b2b8215be2f1de.

sravfeyn committed:
Update index mapping

sravfeyn committed:
make the util a unit and quickcache the unit

sravfeyn commented:
Jenny, I added a quickcache decorator since this seldom changes and is loaded on all view mobile-worker pages.

@snopoke I have updated the mapping to not_analyzed. Since the app index size is small enough, we can just trust that normal deploy would handle the reindex well enough?

sravfeyn committed:
unused import

snopoke reviewed (commented):
@sravfeyn I think prod may need a preindex but other envs should be ok. Any reason you only marked those `string` types as `not_analyzed` and not others?

sravfeyn committed:
update test

sravfeyn committed:
cache it

sravfeyn committed:
update test data

sravfeyn committed:
more fields not_analyzed

sravfeyn commented:
> @sravfeyn I think prod may need a preindex but other envs should be ok. Any reason you only marked those string types as not_analyzed and not others?

Partly, because I wasn't sure if changing other fields would break any existing searches, there's a lot of fields in this mapping which could be set to `not_analyzed` otherwise. I could still try to infer from the field names and do a best guess, but it would still be a guess (which I did now for some fields). My original intention was just to fix the users-page bug without having to deal with reindex on bunch of envs.

sravfeyn commented:
@orangejenny @esoergel @dannyroberts Can any of you guys review the mapping change, please (while Simon is offline).

sravfeyn committed:
do avoid agg

dannyroberts commented:
kicked the one job that failed in the last build due to the recent kafka docker timeout issue

mkangia commented on corehq/apps/app_manager/dbaccessors.py:
```diff
@@ -540,6 +540,21 @@ def get_case_types_from_apps(domain):
+            .run().hits)
+    langs = set()
+    for lang_list in result:
+        langs = langs.union(set(lang_list['langs']))
```
not sure which one is more expensive
```
langs = list(map(lambda x: set(x['langs']), result))
return set().union(*langs)
```
was just curious about this approach, fine to not change.

snopoke replied:
they both have the same cost since they are doing the same thing

mkangia replied:
Cool, i just thought if the single call to `union` with all lists would make a difference.
+ Less lines of code

sravfeyn replied:
Not a hard preference, but I like the current one, since it's slightly more readable (I feel) with no performance difference.

sravfeyn commented:
Running a preindex on prod, as the hqapps index is around 21GB on prod.

sravfeyn commented:
Preindex finished on prod. This can be merged now.
